### ISSUE TITLE
Deep Integration of archivemount

### ISSUE THREAD

**Issue by Naheel-Azawy** [gokcehan/lf#170]
I've integrated archivemount with lf with tiny configs to achieve browsing archives like directories and this was the result:
[https://youtu.be/URt-jH0Bc0Y](https://youtu.be/URt-jH0Bc0Y)
This was added to the integrations wiki too. Even though it looks too much to fit there...
However! This is not enough from my point of view and I think it should be deeply integrated in the source for more seamless usage.
Requirements:

1. Opening an archive would just look opening it and not creating a new directory and going to it.
2. Mount directories junk organized in some hidden place out of the user's eyes.
3. Unmounting should be done at the server's death and not clients as more than one client can be opening the same archive.
4. Show the user that this is an archive and not a normal directory (in the bottom bar?).
5. Whatever else that I've missed.

**Comment by gokcehan** [comment]
@Naheel-Azawy Thanks for adding this to the integrations wiki. I have been using vim to display the content of archive files and open specific files within which provides good enough functionality for my needs so far. I have never heard of archivemount before. It looks like an interesting program though I don't quite understand why it needs to create extra temporary files.

Unfortunately, integrating this to `lf` may not be feasible. First of all, libarchive seems to be a library written in C. We don't have C library dependencies as this would effect our portability and static linking with the requirement of cgo. Second, as far as I know, FUSE filesystems are not possible in windows which we officially support. Lastly, I'm not sure if this would be a good idea at all times. It certainly seems to work for small archives but it can be a problem for larger archive files. Also it seems that it is not possible to write within an archive directory so it can get confusing for users. I'm simply not convinced that having such a seamless experience for archive files is always a good idea. Having this integrated to `lf` seems like a rather opinionated decision.

**Comment by Naheel-Azawy** [comment]
@gokcehan Even though I live in the Emacs land and I'm supposed to start a war with you now but I actually am kind of convinced!
But I still don't find it as a cool idea to use an editor for that and I still think that it would be cool if `lf` can do it. Not with this archivemount probably, but maybe somehow, someone implement something similar in go with no FUSE dependency. Or perhaps an browsing archives gets implemented here specifically for `lf` in the future?
I've changed the title, maybe, in the far future my dreams come true.

**Comment by gokcehan** [comment]
@brualan I don't think it is a good idea. With the standard packages you mentioned, I don't think this is quite possible either without a quite significant effort. Even if we manage to make a prototype, as you say, this would increase the complexity of the code base quite a lot. Also, this would likely be something that would only work for small toy examples and break for those archive files that would not fit into the memory. We would rather not merge something like that to the code base. All in all, I don't see any significant gain from such a feature either. I don't think we need that kind of transparency when working with archive files.

**Comment by usama-makhzoum** [comment]
do we need libarchive? why can't use this https://github.com/mholt/archiver

**Comment by kormat** [comment]
This doesn't fulfill all of the original requirements, but it does allow simple mounting and unmounting of archives.

Features:
- can use archivemount or ratarmount (which is Much faster for large archives)
- uses mountpoints in a tmpdir, so that you can mount archives from directories you don't have write access to
- remembers the source archive that corresponds to a mountpoint
- you can unmount the archive from anywhere in the mountpoint tree
- it doesn't require restarting lf, or using a wrapper script
- after unmounting, it'll switch back to the source archive

I wanted to check that it would be ok to update the [integrations entry](https://github.com/gokcehan/lf/wiki/Integrations#archivemount) with this.

```
set shell bash

cmd archivemount %{{
    if [ ! -e "$f" -o -d "$f" ]; then
        lf -remote "send $id echoerr 'ERROR: not a valid archive'"
        exit 1
    fi
    base=$(basename "$f")
    mntdir="$(mktemp -d --tmpdir lf-am-${base}.XXXX)"
    if type ratarmount &>/dev/null; then
        ratarmount --index-file=":memory:" "$f" "$mntdir"
    else
        archivemount -o readonly "$f" "$mntdir"
    fi
    echo "$f" > "$mntdir.source"
    lf -remote "send $id :cd '$mntdir'; echomsg Mounted $f"                                                            
}}      
map am archivemount                                                                                                    
    
cmd fuseumount %{{
    [ -e "$f" ] || f=.
    mntdir=$(stat -c %m "$f")                                                                                          
    if [ "$mntdir" = "/" ]; then                                                                                       
        lf -remote "send $id echomsg 'mountpoint is /, cowardly doing nothing'"                                        
        exit 1 
    fi
    if ! findmnt -no FSTYPE "$mntdir" | grep -q '^fuse'; then                                                          
        lf -remote "send $id echoerr 'ERROR: not a fuse mountpoint $mntdir'"                                           
        exit 1
    fi  
    cmd="send $id :select '$mntdir'; _do_fuseumount; _rmdir"                                                           
    if [ -e "$mntdir.source" ]; then
        cmd+="; select '$(<${mntdir}.source)'"
        rm "$mntdir.source"                                                                                            
    fi                                                                                                                 
    lf -remote "$cmd"                                                                                                  
}}  
cmd _do_fuseumount %fusermount -u "$f"                                                                                 
cmd _rmdir %rmdir "$f"                                                                                                 
map au fuseumount                                                                                                      
```

**Comment by og900aero** [comment]
100% go solution: [https://github.com/mholt/archives](https://github.com/mholt/archives)

**Comment by joelim-work** [comment]
Using a 3rd party library should make it possible to traverse and list the files in an archive. It might even be possible to support write operations (i.e. adding/renaming/deleting files in the archive).

One thing I'm not sure of however is how to use the `previewer` script to preview files in the archive. The `previewer` is an external script that runs outside of the context of the 3rd party library, which means you would end up having to extract the files anyway.

**Comment by lucas-mior** [comment]
Having transparent compression in a file manager is not that useful considering that filesystems like btrfs already do that. But it is certainly confusing for a user. The way lf works is only listing directories, and work on the files is done by the user/commands. It is a deviation from the way lf has always worked.

I also question the usefulness of it in general. If you downloaded some zip from the internet, why would you want to decompress it every time you have to access it? Why not just decompress it once and have it as a dir/files you can use any tool on, not just compression-aware tools?

**Comment by CatsDeservePets** [comment]
> Having transparent compression in a file manager is not that useful considering that filesystems like btrfs already do that. But it is certainly confusing for a user. The way lf works is only listing directories, and work on the files is done by the user/commands. It is a deviation from the way lf has always worked.

You are talking from a Linux user perspective. Don't forget that we also have many `Windows` users. For them, `btrfs` is not an option. Also, for `Windows` users, browsing archives as if they were normal directories, is expected as it's what `File Explorer` does as well.

> I also question the usefulness of it in general. If you downloaded some zip from the internet, why would you want to decompress it every time you have to access it? Why not just decompress it once and have it as a dir/files you can use any tool on, not just compression-aware tools?

I think this is something really personal that varies depending on your workflow. In my case, I have to download lots of `archives` at work (on Windows, sadly). I like being able to "open" them and select & copy only the files I actually need out of them without the need to extract it first which either fills up the parent directory or creates a new directory which I first have to navigate to. (I bulk-delete the archives later every couple of days or weeks).

We also use archives to store backups in a more efficient way and it's nice being able to see what's inside.


Of course, more efficient workflows might exists but everyone has their own way of doing things.

**Comment by joelim-work** [comment]
The way I see things is that it should be possible to implement this as a virtual directory (similar to #1067), if you try hard enough. But the implementation will likely by non-trivial and also end up having limitations anyway:

- Previewing (and preloading) files won't be possible unless you end up extracting the archive anyway
- Opening the file will have a similar problem, perhaps the file could be extracted into a temp directory
- The `cd` command uses [`os.Chdir`](https://pkg.go.dev/os#Chdir) under the hood, this might need to be enhanced to deal with archives
- Filesystem notifications (the `watch` option) only works on actual directories, so I doubt it will detect changes to an archive.

There might be other limitations I haven't thought of off the top of my head, but in any case I see this kind of feature request as having both pros and cons, and whether or not it is worth implementing is subjective (based on how useful the feature is to a particular user).
